src/main/java/io/vertx/core/impl/DeploymentManager.java

```diff
@@ -79,13 +99,47 @@ public void deployVerticle(Verticle verticle, DeploymentOptions options,
       throw new IllegalArgumentException("Can't specify isolatedClasses for already created verticle");
     }
     ContextImpl currentContext = vertx.getOrCreateContext();
-    doDeploy("java:" + verticle.getClass().getName(), generateDeploymentID(), options, currentContext, currentContext, completionHandler,
-        getCurrentClassLoader(), verticle);
+    ClassLoader cl = getClassLoader(options, currentContext);
+    currentContext.<Set<Verticle>>executeBlocking(fut -> {
+      int nbInstances = options.getInstances();
+      IdentityHashMap<Class<? extends Verticle>, Set<Verticle>> verticles = new IdentityHashMap<>();
+      for (int i = 0; i < nbInstances; i++) {
+        Verticle verticle = verticleSupplier.get();
+        if (verticle == null) {
+          throw new RuntimeException("Supplied verticle is null");
+        }
+        verticles.compute(verticle.getClass(), (k, v) -> {
+          if (v == null) v = new HashSet<>();
+          v.add(verticle);
+          return v;
+        });
+      }
+      if (verticles.size() != 1) {
+        throw new RuntimeException("Supplied verticles are not from the same class");
+      }
+      Set<Verticle> instances = verticles.values().iterator().next();
+      if (instances.size() != nbInstances) {
+        throw new RuntimeException("Same verticle supplied more than once");
+      }
+      fut.complete(instances);
+    }, false, ar -> {
+      if (ar.succeeded()) {
+        Set<Verticle> result = ar.result();
+        Verticle[] verticles = result.toArray(new Verticle[result.size()]);
+        String verticleClass = verticles[0].getClass().getName();
+        doDeploy("java:" + verticleClass, generateDeploymentID(), options, currentContext, currentContext, completionHandler, cl, verticles);
+      } else {
+        completionHandler.handle(Future.failedFuture(ar.cause()));
+      }
+    });
```

Comment:
You can use _distinct()_ because it relies on equals, not identity. We really want different instances